Refactor code for the move to Kubernetes

nikodemas opened the pull request:
This PR together with https://github.com/dmwm/CMSKubernetes/pull/1559 prepares the `udp-collector` for the move to the Kubernetes infrastructure.

FYI @vkuznet @leggerf @brij01

vkuznet commented:
@nikodemas , I doubt that the changes you propose are valid. Let me break down for you the k8s requirements:
- the `/health` end-point should provide status of underlying service. In case of your changes you have UDP server serving UPD requests and separate goroutine (thread) of HTTP server using health end-point. If UDP server experience an issue the HTTP server knows nothing about it. Therefore proposed implementation does not capture the state of health of UDP server
- removing shell script has nothing to do with k8s, I suggest to keep them around for references. You may not need to pack them into a final image.
- you remove code with prometheus metrics which are useful to understand performance of the underlying service. In k8s you can rely on those to know what are the limits required to ru the service.

I suggest to re-factor the code in the following way:
- you may keep your HTTP server goroutine but it should do ping request to UDP server and receive back a reply. Doing this way it will **probe** the status of UDP server
- you may keep code under one main which will run both UDP and HTTP server but you must expose proper ports for them in k8s, one to serve UDP incoming requests, and another to expose HTTP to make health probes
- I suggest to put back prometheus metrics and use them within UDP server. What we should measure is performance of UDP server and not HTTP one. You may create a new UDP request for that, similar how original code is doing ping/pong exchange. A new UDP request will return UDP server prometheus metrics. And you HTTP server will need another end-point `/metrics` which will place UDP request to UDP server to get its metrics.

nikodemas commented:
@vkuznet thanks for your comments. I have adjusted everything accordingly:
- the monitoring service now constantly sends `ping` messages to the udp server and `/health` endpoint checks the time since the last `pong` is received
- Prometheus metrics are added back to the udp server monitor script
- `udp_server` and `udp_server_monitor` are now started from one main since I couldn't start them with a shell script from a distroless image. however, having a shell script doesn't make sense anymore because now everything can be started with a single go executable, so I would like to keep the shell script deleted
- I also updated the version in the GitHub actions as it was failing because of some package mismatch
- after this is merged I should update the `go.mod` and the `main.go` to reference the `github.com/dmwm/udp-collector/` again instead of the local folder

vkuznet commented on README.md:
```diff
@@ -2,67 +2,29 @@
-go build udp_server_monitor.go
-go build udp_client.go
+# build executable
+go build -o udp_collector
```
add here `CGO_ENABLED=0` before go command

vkuznet commented on Makefile:
```diff
@@ -4,34 +4,27 @@ flags=-ldflags="-s -w -X main.version=${VERSION}"
 build:
-	go clean; rm -rf pkg; go build -o udp_server ${flags} udp_server.go
-	go build -o udp_server_monitor ${flags} udp_server_monitor.go
+	go clean; rm -rf pkg; go build -o udp_collector ${flags}
```
for all go build commands add `CGO_ENABLED=0 go build ...` in every build action of Makefile. This way your CI/CD will test the build of static executable which you intend to put on k8s.

vkuznet commented on Makefile:
```diff
@@ -42,4 +35,4 @@ clean:
 
 test1:
-	go test
+	go test
```
For testing, please use `go test -v` or even better `go test -v -race -bench=.` if you write bench tests. The race flag will test go code for racing conditions too.

vkuznet commented on go.mod:
```diff
@@ -1,24 +1,33 @@
-module github.com/dmwm/udp-collector
+module udp-collector
```
There is no need to change module path even if your package is not in github. This is treated as simple string. Very often I create package with these commands:

```
go mod init github.com/..../package
go mod tidy
```
and I suggest that you do the same to rebuild the go.mod and go.sum.

vkuznet commented on go.mod:
```diff
@@ -1,24 +1,33 @@
-module github.com/dmwm/udp-collector
+module udp-collector
 
 go 1.20
```
Depending where you develop your code you may consider using newer version of compiler, current one is 1.23.3. And, if you delete these files and recreate them it will properly construct all dependencies. I perform my development on local machine (macOS) and always keep latest go version as it has bug fixes and more importantly optimizations.
